### PR TITLE
Build custom error pages

### LINKED ISSUE #68: Build custom error pages
@HTTPArchive/developers 

If an unexpected error occurs on the server (5xx) or there's an issue with user input (4xx) we should show an error page with more information about what went wrong.

### PR DESCRIPTION
**Pull request by tanhengyeow**
Fixes #68.

The cases below are included:
1. Status code 400 - Bad Request
2. Status code 404 - Page Not Found
3. Status code 500 - Internal Server Error
4. Status code 502 - Bad Gateway

### CODE REVIEW

**Comment by tanhengyeow** [comment]
> Thanks for working on this @tanhengyeow!
> 
> How might we implement this TODO for which we would pass a custom error message to the error page:
> 
> https://github.com/HTTPArchive/almanac.httparchive.org/blob/49fdcf814f6f078c8cdab0e5797c3e6f923993a1/src/validate.py#L41-L45

@rviscomi Adding the error description in this [line](https://github.com/HTTPArchive/almanac.httparchive.org/pull/77/files#diff-fab5b7e3d7ff2dcbe82a5a0cf622fccbR43) does the trick 😄 

This is an example of the custom error page when the user accesses an unsupported year:
image

File: src/templates/base.html [hunk 1]
```diff
@@ -28,28 +28,30 @@
...
-    </div>
+    {% block content %}
+      <div id="navigation">
+        {% block navigation %}{% endblock %}
+      </div>
+      {% set view_args = request.view_args.copy() %}
+      <select class="language-switcher">
+        <option
+          selected="selected"
+          value={{ url_for(request.endpoint, **(view_args.update(lang=self.lang()[0:2]) or view_args)) }}
```
Comment (rviscomi) [inline src/templates/base.html]: FYI just merged #69 so this is outdated now.

Comment (rviscomi) [reply src/templates/base.html]: Otherwise this looks good!

**Comment by rviscomi** [comment]
Looks good. Curious why you made the `if supported_languages is iterable` change. Should always be a list or tuple, right?

**Comment by tanhengyeow** [comment]
> Looks good. Curious why you made the `if supported_languages is iterable` change. Should always be a list or tuple, right?

So when an unsupported language is requested by the user, this [line](https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/src/main.py#L17) returns a non-list/tuple, resulting in a `_Language object not iterable` error. That's the reason why I added that check 😄

**Comment by tanhengyeow** [comment]
> > Looks good. Curious why you made the `if supported_languages is iterable` change. Should always be a list or tuple, right?
> 
> So when an unsupported language is requested by the user, this [line](https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/src/main.py#L17) returns a non-list/tuple, resulting in a `_Language object not iterable` error. That's the reason why I added that check 😄

@rviscomi Come to think of it, should we do this in the backend instead? For e.g. return an empty list/tuple by default if we can't find the supported year. That way we can be sure that it is iterable on the frontend. But even so, we still have to check if `supported_languages` is empty because we do not want to trigger the for loop if `supported_languages` is an empty list/tuple.

**Comment by rviscomi** [comment]
Yes I agree it should be addressed on the backend but I didn't want it to hold up this PR.

**Pull request merged**